### DISCUSSION TITLE
Building cython extensions using only pyproject.toml (no setup.py)

### DISCUSSION THREAD

**Discussion by rusty-electron** [pypa/setuptools#4154] [Q&A] [answered]
I am working on a python project that requires the use of cython extension modules. As of now, I am using both a `pyproject.toml` and a `setup.py` file. The toml file contains the package metadata, entry points, requirements, etc., and the `setup.py` only contains the build specifications for the cython file. Because of this, I wanted to know if it is possible to get rid of the `setup.py` completely and rely only on the `pyproject.toml` file. I looked up about it on links like [this](https://setuptools.pypa.io/en/latest/userguide/ext_modules.html) and [this](https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html) but couldn't find an answer. So is it possible to just have a `pyproject.toml` file and still be able to build the extension modules?

**Comment by abravalheri** [comment, marked as answer]
It is current not possible to configure `ext_modules` in `pyproject.toml`. For the time being you need both files.

**Comment by webknjaz** [comment]
If you're feeling adventurous, check out what I did @ https://github.com/aio-libs/yarl/tree/a3df0aa/packaging. But this is a very non-standard approach that involves monkey-patching, so it's rather dangerous. Do not try at home!

**Reply by rusty-electron** [reply]
Interesting! I'll take a look.

**Comment by stanislawmorawski-digica** [comment]
It's possible with `setuptools > 74.1` and although for now support is experimental I've been able to use it without any hiccups. Here's how:

1. In `pyproject.toml` add `"cython"` to `[build-system].requires`:
 ```
 [build-system]
requires = ["setuptools", "cython"]
```
2. In `pyproject.toml` add your extenstion configurations to `[tool.setuptools].ext-modules`:
```
[tool.setuptools]
ext-modules = [
  {name = "example", sources = ["example.pyx"]} # You can also specify all the usual options like language or include_dirs 
]
```

3. Run `pip install .` (or `pip install -e .`) to trigger the build.  


Edit: See the note in [this part](https://cython.readthedocs.io/en/latest/src/userguide/source_files_and_compilation.html#basic-setup-py) of cython docs

**Reply by AndrewAnnex** [reply]
@stanislawmorawski-digica do you happen to know how to properly set include_dirs with numpy headers? I have a setup.py I am attempting to transition which uses `numpy.get_include()` to find them

**Reply by stanislawmorawski-digica** [reply]
@AndrewAnnex I don't think a clean solution to set them dynamically is possible as of today
